### PR TITLE
perf: Tile 8×8 covariance matrix multiplication

### PR DESCRIPTION
**Pull request by stephenswat**
Currently, we are multiplying an 8x8 covariance matrix with an 8x8 transport matrix, and we see that Eigen is failing to optimize this properly, because it is calling a generalized GEMM method rather than an optimized small matrix method. In order to resolve this, we change the code to use a tiled multiplication method which splits the matrices into 4x4 sub-matrices which can be multiplied and added to achieve the desired effect. This has two advantages:

  1. It allows Eigen to use its hand-rolled optimized 4x4 matrix multiplication methods.
  2. It allows us to perform some trickery with matrix identities to reduce the number of floating point operations.

### CODE REVIEW

**Comment by stephenswat** [comment]
With both this and #1183, my performance for the Eigen stepper becomes 36.595292 ms/event on the generic propagation example, while the performance for the ATLAS stepper is 29.353479 ms/event. Getting closer!

File: Core/include/Acts/Propagator/EigenStepper.ipp [hunk 1]
```diff
@@ -233,8 +233,50 @@ Acts::Result<double> Acts::EigenStepper<E, A>::step(
...
+    // K₁₁ = I₈  * I₈  + D₁₂ * 0₈  = I₈
+    // K₁₂ = I₈  * J₁₂ + D₁₂ * J₂₂ = J₁₂ + D₁₂ * J₂₂
+    // K₂₁ = 0₈  * I₈  + D₂₂ * 0₈  = 0₈
+    // K₂₂ = 0₈  * J₁₂ + D₂₂ * J₂₂ = D₂₂ * J₂₂
+    //
+    // Furthermore, we're constructing K in place of J, and since
+    // K₁₁ = I₈ = J₁₁ and K₂₁ = 0₈ = D₂₁, we don't actually need to touch those
+    // sub-matrices at all!
+    if ((D.topLeftCorner<4, 4>().isIdentity()) &&
+        (D.bottomLeftCorner<4, 4>().isZero()) &&
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/EigenStepper.ipp:264]: Just curious: these checks are likely much lower overhead than calling into GEMM, right?

**Comment by stephenswat** [comment]
Miffed why this fails to be honest.

**Comment by stephenswat** [comment]
Okay, I really think this is just harmless numerical errors causing hash changes.

**Comment by paulgessinger** [comment]
There's still a couple of hash differences in the CI.

**Comment by stephenswat** [comment]
Updated, let's see if this works. Could someone give a cursory check of the physmon plots to see if everything still makes sense? It looks good to me but I am not intimately familiar with how those plots should look.

**Comment by paulgessinger** [comment]
I think they're from #1216, so don't worry about that.

**Comment by stephenswat** [comment]
@paulgessinger passes now.

**Comment by paulgessinger** [comment]
I'll have one more look at the physmon output, and then we merge!

**Comment by paulgessinger** [comment]
I don't quite understand why the physmon artifact doesn't show up in the CI.

**Comment by paulgessinger** [comment]
imageimage
imageimage

Ok so these are the physmon plots that fail comparisons. What's slightly worrisome is the nMeasurements and total number of states, where we lose about 2 in the left most bins. Fully seeded tracking efficiency seems unchanged, but truth estimated tracking efficiency changes: 

imageimage

**Comment by asalzburger** [comment]
Probably something worth discussing today.

**Comment by stephenswat** [comment]
I've changed the Eigen tolerance for determining whether a matrix is the identity or zero matrix to 10<sup>-8</sup>. Let's see if that fixes things.

**Comment by paulgessinger** [comment]
This still changes the number of measurements at negative eta for full seeding:

image

**Comment by asalzburger** [comment]
Does this mean we do not want to merge that in?

**Comment by stephenswat** [comment]
I'll start on a more robust way of doing this.

**Comment by paulgessinger** [comment]
Can we re-evaluate this?

**Comment by asalzburger** [comment]
Retriggering CI here and merge main in.

**Comment by stephenswat** [comment]
Some of these plots still stink. :sob: 

index

**Comment by paulgessinger** [comment]
That's indeed still strange. The track level plots seem identical to me, so where does the difference come from? Propagation by the vertexing itself?

**Comment by stephenswat** [comment]
It's interesting that the bins widths are not the same, i.e. the bins between the reference and monitored histograms do not overlap.

**Comment by paulgessinger** [comment]
Oh you're right. I'm not constraining the range for these plots. Should be an easy fix though.

File: Core/include/Acts/Propagator/EigenStepper.ipp [hunk 2]
```diff
@@ -238,8 +238,41 @@ Acts::Result<double> Acts::EigenStepper<E, A>::step(
...
+    //
+    // K₁₁ = I₈  * I₈  + D₁₂ * 0₈  = I₈
+    // K₁₂ = I₈  * J₁₂ + D₁₂ * J₂₂ = J₁₂ + D₁₂ * J₂₂
+    // K₂₁ = 0₈  * I₈  + D₂₂ * 0₈  = 0₈
+    // K₂₂ = 0₈  * J₁₂ + D₂₂ * J₂₂ = D₂₂ * J₂₂
+    //
+    // Furthermore, we're constructing K in place of J, and since
+    // K₁₁ = I₈ = J₁₁ and K₂₁ = 0₈ = D₂₁, we don't actually need to touch those
+    // sub-matrices at all!
+    if ((D.topLeftCorner<4, 4>().isIdentity(1e-8)) &&
```
Comment (andiwand) [inline Core/include/Acts/Propagator/EigenStepper.ipp:258]: what happens when we lower those thresholds further?

**Review by andiwand — COMMENTED** [review]
just checked the physmon output and it looks like only some random vertexing comparisons fail

**Comment by paulgessinger** [comment]
image

**Comment by paulgessinger** [comment]
How about we try again!

**Comment by paulgessinger** [comment]
The physmon coverage and configuration is much more robust than it was a year or even half a year ago. I don't see any significant changes, other than the ROOT file hashes, everything is green.

Should we just merge this at this stage?

File: Core/include/Acts/Propagator/EigenStepper.ipp [hunk 3]
```diff
@@ -233,8 +233,41 @@ Acts::Result<double> Acts::EigenStepper<E, A>::step(
...
+        (D.bottomLeftCorner<4, 4>().isZero(1e-8)) &&
+        (state.stepping.jacTransport.template topLeftCorner<4, 4>().isIdentity(
+            1e-8)) &&
+        (state.stepping.jacTransport.template bottomLeftCorner<4, 4>().isZero(
+            1e-8))) {
```
Comment (andiwand) [inline Core/include/Acts/Propagator/EigenStepper.ipp:258]: I am a bit worried about the hardcoded thresholds

Comment (paulgessinger) [reply Core/include/Acts/Propagator/EigenStepper.ipp:258]: We can give it another try without them.

**Comment by stephenswat** [comment]
image

File: Core/include/Acts/Propagator/EigenStepper.ipp [hunk 4]
```diff
@@ -233,8 +233,41 @@ Acts::Result<double> Acts::EigenStepper<E, A>::step(
...
+      state.stepping.jacTransport.template topRightCorner<4, 4>() +=
+          D.topRightCorner<4, 4>() *
+          state.stepping.jacTransport.template bottomRightCorner<4, 4>();
+      state.stepping.jacTransport.template bottomRightCorner<4, 4>() =
+          (D.bottomRightCorner<4, 4>() *
+           state.stepping.jacTransport.template bottomRightCorner<4, 4>())
+              .eval();
```
Comment (andiwand) [inline Core/include/Acts/Propagator/EigenStepper.ipp:265]:
```suggestion
    if ((D.topLeftCorner<4, 4>().isIdentity()) &&
        (D.bottomLeftCorner<4, 4>().isZero()) &&
        (state.stepping.jacTransport.template topLeftCorner<4, 4>().isIdentity()) &&
        (state.stepping.jacTransport.template bottomLeftCorner<4, 4>().isZero())) {
      state.stepping.jacTransport.template topRightCorner<4, 4>() +=
          D.topRightCorner<4, 4>() *
          state.stepping.jacTransport.template bottomRightCorner<4, 4>();
      state.stepping.jacTransport.template bottomRightCorner<4, 4>() =
          (D.bottomRightCorner<4, 4>() *
           state.stepping.jacTransport.template bottomRightCorner<4, 4>())
              .eval();
```

**Comment by andiwand** [comment]
looks like the physmon diff is gone

**Comment by andiwand** [comment]
physmon is passing. results here https://github.com/acts-project/acts/pull/1181#issuecomment-1305680691

hashes are changing which means it does at least something 😄 

should be put this in @paulgessinger ?

**Comment by andiwand** [comment]
quick perf measures with https://github.com/andiwand/cern-scripts/blob/main/tmp/full_chain_perf.py

fatras

on average `x 0.98`

image

ckf

on average `x 0.90`

image

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [b83164ac0986ef2baf0e59a90341d45ec2290b8e]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679057)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679056)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679055)||
|:green_circle: | [test_ActsConversionWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679054)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679053)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679052)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679051)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679050)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679049)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679048)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679047)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679046)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679045)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679044)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679043)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679042)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679041)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679040)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679039)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679038)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679037)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679036)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679035)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679034) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_data18_13TeV_1000evt/pulls/1181/)|
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36679033) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_ttbarPU40_reco/pulls/1181/)|

**Review by paulgessinger — APPROVED** [review]
All green, let's go!

**Pull request merged**